Signal/src/ViewControllers/GifPicker/GifPickerCell.swift

```diff
@@ -0,0 +1,161 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+
+class GifPickerCell: UICollectionViewCell {
+    let TAG = "[GifPickerCell]"
+
+    // MARK: Properties
+
+    var imageInfo: GiphyImageInfo? {
+        didSet {
+            AssertIsOnMainThread()
+
+            ensureLoad()
+        }
+    }
+
+    var shouldLoad = false {
+        didSet {
+            AssertIsOnMainThread()
+
+            ensureLoad()
+        }
+    }
+
+    var stillAssetRequest: GiphyAssetRequest?
+    var stillAsset: GiphyAsset?
+    var fullAssetRequest: GiphyAssetRequest?
+    var fullAsset: GiphyAsset?
+    var imageView: YYAnimatedImageView?
+
+    // MARK: Initializers
+
+    @available(*, unavailable, message:"use other constructor instead.")
+    required init?(coder aDecoder: NSCoder) {
+        super.init(coder: aDecoder)
+        owsFail("\(self.TAG) invalid constructor")
+    }
+
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+    }
+
+    override func prepareForReuse() {
+        super.prepareForReuse()
+
+        imageInfo = nil
+        shouldLoad = false
+        stillAsset = nil
+        stillAssetRequest?.cancel()
+        stillAssetRequest = nil
+        fullAsset = nil
+        fullAssetRequest?.cancel()
+        fullAssetRequest = nil
+        imageView?.removeFromSuperview()
+        imageView = nil
+    }
+
+    private func clearStillAssetRequest() {
+        stillAssetRequest?.cancel()
+        stillAssetRequest = nil
+    }
+
+    private func clearFullAssetRequest() {
+        fullAssetRequest?.cancel()
+        fullAssetRequest = nil
+    }
+
+    private func clearAssetRequest() {
+        clearStillAssetRequest()
+        clearFullAssetRequest()
+    }
+
+    private func ensureLoad() {
+        guard shouldLoad else {
+            clearAssetRequest()
+            return
+        }
+        guard let imageInfo = imageInfo else {
+            clearAssetRequest()
+            return
+        }
+        guard self.fullAsset == nil else {
+            return
+        }
+        guard let fullRendition = imageInfo.pickGifRendition() else {
```

Comment:
The Giphy API returns a slew of "renditions" for a given image.  It's critical that we carefully "pick" the best rendition to use.